SEM-lab1/Assets/Scripts/PlayerController.cs

```diff
@@ -1,34 +1,79 @@
 ﻿using System.Collections;
 using System.Collections.Generic;
 using UnityEngine;
+using UnityEngine.SceneManagement;
 
 public class PlayerController : MonoBehaviour
 {
     GameController _gameController;
-    public float speed = 5.0f;
+    public float rotateSpeed = 5.0f;
+    public float powerMin = 5f;             
+    public float powerMax = 500f;              
+    public float chargeTime = 5f;     
+    private float _chargeTimeCurr = 0f;
+    private bool _isCharging = false;       
+    private bool _spaceUp = true;
+    private bool _fired = false;
+    private Quaternion _originalRotation;
 
     // Start is called before the first frame update
     void Start()
     {
         _gameController = GameObject.FindGameObjectWithTag("GameController").GetComponent<GameController>();
+        this._originalRotation = this.transform.rotation; //saves initial rotation for refiring so that the ball is not at a strange angle due to rolling
     }
 
     // Update is called once per frame
     void Update()
     {
+        if(!_fired) //can only rotate if ball is not moving (not being fired)
+        {
+            transform.Rotate(-Input.GetAxis("Vertical") * rotateSpeed, Input.GetAxis("Horizontal") * rotateSpeed, 0.0f);
+            //rotates the ball up/down, left/right on its axis
+        }
 
-        float horizontalMovement = Input.GetAxis("Horizontal");
-        float verticalMovement = Input.GetAxis("Vertical");
-        Debug.Log("Input: " + horizontalMovement + " " + verticalMovement);
+        if (Input.GetKeyDown(KeyCode.Space) && _spaceUp && _fired == false)
+        {
+            _chargeTimeCurr = 0f; //the longer space is pressed = more power
+            _isCharging = true;
+            Debug.Log("Charge time: " + _chargeTimeCurr);
+        }
+
+        if (_isCharging && Input.GetKeyUp(KeyCode.Space) && _fired == false)
+        {
+            _spaceUp = true;
+            float power = Mathf.Lerp(powerMin, powerMax, _chargeTimeCurr / chargeTime);
+            //interpolates between min power, max power over the max charge time
+            Debug.Log("Power: " + power);
+            _fired = true;
+            HitBall(power);
+        }
+
+        if (_isCharging)
+        {
+            _chargeTimeCurr += Time.deltaTime; //increases charge over time, when space is pressed
+        }
 
         Rigidbody r = GetComponent<Rigidbody>();
-        r.velocity = new Vector3(horizontalMovement * speed, 0.0f, verticalMovement * speed);
+        float movementSpeed = r.velocity.magnitude;
+        if (movementSpeed < 0.05 && _fired) //checks if the ball is still moving, ie has a magnitude
```

Comment:
Why 0.05 for the movementSpeed?

Reply:
Mostly derived from trial and error. If movementSpeed == 0 then the ball appears to have stopped long before it has 'technically' stopped which could be frustrating for the player. The actual value is subject to change depending on feedback.